tsdb/db.go

```diff
@@ -1476,32 +1476,42 @@ func (db *DB) CleanTombstones() (err error) {
 	start := time.Now()
 	defer db.metrics.tombCleanTimer.Observe(time.Since(start).Seconds())
 
-	newUIDs := []ulid.ULID{}
-	defer func() {
-		// If any error is caused, we need to delete all the new directory created.
-		if err != nil {
-			for _, uid := range newUIDs {
-				dir := filepath.Join(db.Dir(), uid.String())
-				if err := os.RemoveAll(dir); err != nil {
-					level.Error(db.logger).Log("msg", "failed to delete block after failed `CleanTombstones`", "dir", dir, "err", err)
-				}
-			}
-		}
-	}()
+	var (
+		cleanUpCompleted = false
+		blocks           []*Block
+	)
 
-	db.mtx.RLock()
-	blocks := db.blocks[:]
-	db.mtx.RUnlock()
+	// Repeat cleanup until there is no tombstones left
+	for !cleanUpCompleted {
+		cleanUpCompleted = true
 
-	for _, b := range blocks {
-		if uid, er := b.CleanTombstones(db.Dir(), db.compactor); er != nil {
-			err = errors.Wrapf(er, "clean tombstones: %s", b.Dir())
-			return err
-		} else if uid != nil { // New block was created.
-			newUIDs = append(newUIDs, *uid)
+		blocks = db.Blocks()
+		for _, b := range blocks {
+			if b.tombstones.Total() > 0 {
+				cleanUpCompleted = false
+
+				if uid, er := b.CleanTombstones(db.Dir(), db.compactor); er != nil {
+					return errors.Wrapf(er, "clean tombstones: %s", b.Dir())
+				} else if uid != nil { // New block was created.
+					if err = db.reload(); err != nil {
+						// db.reload will try to delete old block
```

Comment:
It's probably best to just fail here, rather than messing around further after an error has occured.

Reply:
Please correct me if I'm wrong, but b.CleanTombstones will create a new block, reload will try to delete the old one. If it fails to delete and we leave the two blocks there, we will have overlapping timeseries.

On this part of the code, I'm basically refactoring the deferred function that deleted all new blocks after a failed reload. Since we are deleting one by one now, we can do this validation one by one as well

Reply:
After another read on the reload() function, I've noticed that parent(old) blocks are [marked as deletable](https://github.com/prometheus/prometheus/blob/d6fa7f72a627564966e9429d3c106685fbb7c340/tsdb/db.go#L922-L935) already.

I was wondering if, after failing to delete an old block, it is really a problem to not delete the new compacted ones (today the new ones are being deleted after failing to delete the old ones). Won't the reload() function be called multiple times on other operations and, thus, there will be several other retries to delete the old blocks?

Reply:
Exactly, there's no need to do much beyond returning an error. If this fails we're already in a bad state.

Reply:
After a really dirty debugging and a lot of "fmt.Println" 🥵 I've finally found out that tests were failing after I removed this error handling (at Linux at least). A new compacted block is created, it then fails to delete the old one due to overlapping block sequence. Currently, tests expect that the new compacted block is deleted after a failed reload. 

--

So, just to make things a little bit clearer for me... Should I really leave new blocks there after failing to remove old ones, then change tests accordingly? Or should I keep this behavior?

Personally, I would prefer to keep this behavior, I don't see why leaving both the new compacted block and its parent there... I think one of the two should be deleted.

WDYT @brian-brazil @codesome @bwplotka?